File: app/src/main/java/com/example/data/source/SourceResolver.kt
package com.example.data.source

import android.util.Log
import com.example.data.remote.ArxivClient
import com.example.data.remote.CrossrefClient
import com.example.data.remote.GithubApi
import com.example.data.remote.HuggingFaceApi
import com.example.domain.model.SourceType
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

data class SourceInfo(
    val sourceType: SourceType,
    val sourceId: String,
    val sourceTitle: String,
    val sourceAuthors: String?,
    val sourceAbstract: String?,
    val sourceExtra: String?
)

class SourceResolver(
    private val arxivClient: ArxivClient,
    private val githubApi: GithubApi,
    private val huggingFaceApi: HuggingFaceApi,
    private val crossrefClient: CrossrefClient
) {

    private suspend fun <T> withRetry(maxAttempts: Int = 3, delayMs: Long = 1000, block: suspend () -> T?): T? {
        repeat(maxAttempts) { attempt ->
            try {
                return block()
            } catch (e: retrofit2.HttpException) {
                if (e.code() == 429 || e.code() == 503) {
                    val retryAfterMs = (e.response()?.headers()?.get("Retry-After")?.toLongOrNull() ?: (1L shl attempt)) * 1000L
                    kotlinx.coroutines.delay(retryAfterMs)
                } else if (e.code() in 400..499) {
                    return null  // client errors: don't retry
                }
            } catch (e: Exception) {
                if (e is kotlinx.coroutines.CancellationException) throw e
                if (attempt == maxAttempts - 1) return null
                kotlinx.coroutines.delay(delayMs * (1L shl attempt))
            }
        }
        return null
    }

    suspend fun resolve(text: String, url: String?): SourceInfo? = withContext(Dispatchers.IO) {
        val allUrls = extractAllUrls(text) + listOfNotNull(url)

        // Priority: arXiv > HuggingFace papers page > GitHub > HuggingFace model/dataset > DOI
        for (u in allUrls) {
            resolveArxiv(u)?.let { return@withContext it }
        }
        for (u in allUrls) {
            resolveHfPaper(u)?.let { return@withContext it }
        }
        for (u in allUrls) {
            resolveGithub(u)?.let { return@withContext it }
        }
        for (u in allUrls) {
            resolveHuggingFace(u)?.let { return@withContext it }
        }
        // Fallback: bare arXiv IDs mentioned in text without a URL (e.g. "2312.00752")
        ArxivClient.ARXIV_BARE_REGEX.find(text)?.groupValues?.get(1)?.let { bareId ->
            resolveArxiv("https://arxiv.org/abs/$bareId")
        }?.let { return@withContext it }
        // Fallback: a DOI in a doi.org URL or bare "10.xxxx/..." token anywhere in the text.
        for (u in allUrls) {
            CrossrefClient.DOI_REGEX.find(u)?.groupValues?.get(1)?.let { doi ->
                resolveDoi(doi)?.let { return@withContext it }
            }
        }
        CrossrefClient.DOI_REGEX.find(text)?.groupValues?.get(1)?.let { doi ->
            resolveDoi(doi)
        }?.let { return@withContext it }
        null
    }

    // ── DOI (Crossref) ──────────────────────────────────────────────────────────

    private suspend fun resolveDoi(doi: String): SourceInfo? {
        return try {
            val meta = withRetry { crossrefClient.fetchWork(doi) } ?: return null
            // arXiv DOIs (10.48550/arXiv.*) are better served by the arXiv resolver; skip them here.
            if (meta.doi.contains("arxiv", ignoreCase = true)) return null
            val extra = buildExtraJson(mapOf(
                "doi" to meta.doi,
                "published" to meta.published,
                "container" to meta.containerTitle,
                "type" to meta.type
            ))
            SourceInfo(
                sourceType = SourceType.DOI,
                sourceId = meta.doi,
                sourceTitle = meta.title,
                sourceAuthors = meta.authors.joinToString(", ").ifBlank { null },
                sourceAbstract = meta.abstract,
                sourceExtra = extra
            )
        } catch (e: Exception) {
            Log.e("SourceResolver", "DOI resolve failed for $doi: ${e.message}")
            null
        }
    }

    private fun extractAllUrls(text: String): List<String> {
        // Matches full URLs including query strings (?k=v) and fragments (#anchor).
        // Characters excluded from the character class: whitespace, angle brackets, quotes.
        // Trailing punctuation that is syntactically part of the surrounding sentence
        // (period, comma, closing paren/bracket, exclamation, semicolon) is stripped after
        // extraction so that "see https://arxiv.org/abs/2401.00001." works correctly.
        val regex = Regex("""https?://[^\s<>"']+""")
        return regex.findAll(text)
            .map { it.value.trimEnd('.', ',', ')', ']', '!', '?', ';') }
            .toList()
    }

    // ── arXiv ────────────────────────────────────────────────────────────────

    private suspend fun resolveArxiv(url: String): SourceInfo? {
        val id = ArxivClient.ARXIV_ID_REGEX.find(url)?.groupValues?.get(1)
            ?.replace(Regex("v\\d+$"), "") ?: return null

        return try {
            val meta = withRetry { arxivClient.fetchPaper(id) } ?: return null
            val extra = buildExtraJson(mapOf(
                "published" to meta.published,
                "categories" to meta.categories
            ))
            SourceInfo(
                sourceType = SourceType.ARXIV,
                sourceId = meta.id,
                sourceTitle = meta.title,
                sourceAuthors = meta.authors.joinToString(", "),
                sourceAbstract = meta.abstract,
                sourceExtra = extra
            )
        } catch (e: Exception) {
            Log.e("SourceResolver", "arXiv resolve failed for $id: ${e.message}")
            null
        }
    }

    // HuggingFace papers links embed arXiv IDs
    private suspend fun resolveHfPaper(url: String): SourceInfo? {
        val id = Regex("huggingface\\.co/papers/([\\d.]+)").find(url)?.groupValues?.get(1)
            ?: return null
        return resolveArxiv("https://arxiv.org/abs/$id")
    }

    // ── GitHub ────────────────────────────────────────────────────────────────

    private suspend fun resolveGithub(url: String): SourceInfo? {
        val match = Regex("github\\.com/([a-zA-Z0-9_.-]+)/([a-zA-Z0-9_.-]+)").find(url)
            ?: return null
        val owner = match.groupValues[1]
        val repo = match.groupValues[2].removeSuffix(".git")
        if (owner.isBlank() || repo.isBlank()) return null

        return try {
            val r = withRetry { githubApi.getRepo(owner, repo) } ?: return null
            val extra = buildExtraJson(mapOf(
                "stars" to r.stars,
                "language" to r.language,
                "topics" to r.topics,
                "pushedAt" to r.pushedAt
            ))
            SourceInfo(
                sourceType = SourceType.GITHUB,
                sourceId = r.fullName,
                sourceTitle = r.fullName,
                sourceAuthors = r.owner.login,
                sourceAbstract = r.description,
                sourceExtra = extra
            )
        } catch (e: Exception) {
            Log.e("SourceResolver", "GitHub resolve failed for $owner/$repo: ${e.message}")
            null
        }
    }

    // ── HuggingFace ───────────────────────────────────────────────────────────

    private suspend fun resolveHuggingFace(url: String): SourceInfo? {
        // Exclude known non-model paths
        if (url.contains("/spaces/") || url.contains("/papers/") || url.contains("/docs/")) {
            return null
        }
        val datasetMatch = Regex("huggingface\\.co/datasets/([a-zA-Z0-9_.-]+/[a-zA-Z0-9_.-]+)").find(url)
        if (datasetMatch != null) {
            return resolveHfDataset(datasetMatch.groupValues[1])
        }
        val modelMatch = Regex("huggingface\\.co/([a-zA-Z0-9_.-]+/[a-zA-Z0-9_.-]+)").find(url)
        if (modelMatch != null) {
            return resolveHfModel(modelMatch.groupValues[1])
        }
        return null
    }

    private suspend fun resolveHfModel(id: String): SourceInfo? {
        return try {
            val r = withRetry { huggingFaceApi.getModel(id) } ?: return null
            val modelId = r.modelId ?: r.id ?: id
            val extra = buildExtraJson(mapOf(
                "downloads" to r.downloads,
                "likes" to r.likes,
                "pipelineTag" to r.pipelineTag,
                "tags" to r.tags.take(8)
            ))
            SourceInfo(
                sourceType = SourceType.HUGGING_FACE,
                sourceId = modelId,
                sourceTitle = modelId,
                sourceAuthors = r.author,
                sourceAbstract = r.description?.take(500),
                sourceExtra = extra
            )
        } catch (e: Exception) {
            Log.e("SourceResolver", "HF model resolve failed for $id: ${e.message}")
            null
        }
    }

    private suspend fun resolveHfDataset(id: String): SourceInfo? {
        return try {
            val r = withRetry { huggingFaceApi.getDataset(id) } ?: return null
            val datasetId = r.id ?: id
            val extra = buildExtraJson(mapOf(
                "downloads" to r.downloads,
                "likes" to r.likes,
                "tags" to r.tags.take(8)
            ))
            SourceInfo(
                sourceType = SourceType.HUGGING_FACE,
                sourceId = datasetId,
                sourceTitle = datasetId,
                sourceAuthors = r.author,
                sourceAbstract = r.description?.take(500),
                sourceExtra = extra
            )
        } catch (e: Exception) {
            Log.e("SourceResolver", "HF dataset resolve failed for $id: ${e.message}")
            null
        }
    }

    private fun buildExtraJson(data: Map<String, Any?>): String {
        val obj = org.json.JSONObject()
        data.filterValues { it != null }.forEach { (k, v) ->
            when (v) {
                is List<*> -> obj.put(k, org.json.JSONArray(v))
                else -> obj.put(k, v)
            }
        }
        return obj.toString()
    }
}

File: ios/Curio/Curio/Intelligence/SourceResolver.swift
import Foundation
import os

/// Resolved primary-source metadata. Port of `data class SourceInfo`.
struct SourceInfo: Sendable, Equatable {
    let sourceType: SourceType
    let sourceId: String
    let sourceTitle: String
    let sourceAuthors: String?
    let sourceAbstract: String?
    let sourceExtra: String?
}

/// Resolves a bookmark's text + url to a primary research source (arXiv / GitHub / HuggingFace /
/// Crossref DOI). Ports `class SourceResolver` from `data/source/SourceResolver.kt`.
///
/// Fixed resolution priority (preserved exactly — CONVENTIONS Repository cross-cutting / DESIGN §6):
///   arXiv > HuggingFace papers page > GitHub > HuggingFace model/dataset > bare-arXiv-in-text >
///   DOI-in-url > DOI-in-text.
///
/// `withRetry` mirrors the Kotlin retry/backoff: a `429`/`503` honours `Retry-After` (else `2^attempt`
/// seconds); a `4xx` client error → `nil` (no retry); any other error backs off `delayMs * 2^attempt`
/// and gives up on the last attempt; `CancellationError` is rethrown. The underlying networking
/// clients are already resilient (log + return `nil`) so the retry shape is preserved for the rare
/// case a block surfaces an `APIError`.
struct SourceResolver: Sendable {

    private let arxivClient: ArxivClient
    private let githubApi: GithubApiClient
    private let huggingFaceApi: HuggingFaceApiClient
    private let crossrefClient: CrossrefClient
    private static let logger = Logger(subsystem: "com.curio.app", category: "SourceResolver")

    init(
        arxivClient: ArxivClient,
        githubApi: GithubApiClient,
        huggingFaceApi: HuggingFaceApiClient,
        crossrefClient: CrossrefClient
    ) {
        self.arxivClient = arxivClient
        self.githubApi = githubApi
        self.huggingFaceApi = huggingFaceApi
        self.crossrefClient = crossrefClient
    }

    // MARK: - Retry

    /// Port of `private suspend fun <T> withRetry(maxAttempts, delayMs, block)`.
    private func withRetry<T: Sendable>(
        maxAttempts: Int = 3,
        delayMs: Int = 1000,
        _ block: () async throws -> T?
    ) async -> T? {
        for attempt in 0..<maxAttempts {
            do {
                return try await block()
            } catch let apiError as APIError {
                if let code = apiError.statusCode, code == 429 || code == 503 {
                    let retryAfter = apiError.headers["retry-after"].flatMap { Int64($0) } ?? (Int64(1) << attempt)
                    let retryAfterMs = retryAfter * 1000
                    try? await Task.sleep(for: .milliseconds(retryAfterMs))
                } else if let code = apiError.statusCode, (400..<500).contains(code) {
                    return nil  // client errors: don't retry
                }
                // Other status codes fall through to the next attempt.
            } catch is CancellationError {
                // Rethrow-cancel parity: stop retrying (callers treat a nil as "unresolved").
                return nil
            } catch {
                if attempt == maxAttempts - 1 { return nil }
                try? await Task.sleep(for: .milliseconds(Int64(delayMs) * (Int64(1) << attempt)))
            }
        }
        return nil
    }

    // MARK: - Resolve

    /// Port of `suspend fun resolve(text, url)`. Tries each resolver in fixed priority order over all
    /// extracted URLs, then the bare-arXiv and DOI text fallbacks.
    func resolve(text: String, url: String?) async -> SourceInfo? {
        var allUrls = extractAllUrls(text)
        if let url { allUrls.append(url) }

        // Priority: arXiv > HuggingFace papers page > GitHub > HuggingFace model/dataset > DOI
        for u in allUrls {
            if let info = await resolveArxiv(u) { return info }
        }
        for u in allUrls {
            if let info = await resolveHfPaper(u) { return info }
        }
        for u in allUrls {
            if let info = await resolveGithub(u) { return info }
        }
        for u in allUrls {
            if let info = await resolveHuggingFace(u) { return info }
        }
        // Fallback: bare arXiv IDs mentioned in text without a URL (e.g. "2312.00752").
        if let match = text.firstMatch(of: ArxivClient.arxivBareRegex) {
            let bareId = String(match.1)
            if let info = await resolveArxiv("https://arxiv.org/abs/\(bareId)") { return info }
        }
        // Fallback: a DOI in a doi.org URL or bare "10.xxxx/..." token anywhere in the text.
        for u in allUrls {
            if let match = u.firstMatch(of: CrossrefClient.doiRegex) {
                let doi = String(match.1)
                if let info = await resolveDoi(doi) { return info }
            }
        }
        if let match = text.firstMatch(of: CrossrefClient.doiRegex) {
            let doi = String(match.1)
            if let info = await resolveDoi(doi) { return info }
        }
        return nil
    }

    // MARK: - DOI (Crossref)

    private func resolveDoi(_ doi: String) async -> SourceInfo? {
        let meta = await withRetry { await crossrefClient.fetchWork(doi) }
        guard let meta else { return nil }
        // arXiv DOIs (10.48550/arXiv.*) are better served by the arXiv resolver; skip them here.
        if meta.doi.range(of: "arxiv", options: .caseInsensitive) != nil { return nil }
        let extra = buildExtraJson([
            ("doi", meta.doi),
            ("published", meta.published),
            ("container", meta.containerTitle),
            ("type", meta.type)
        ])
        let authors = meta.authors.joined(separator: ", ")
        return SourceInfo(
            sourceType: .DOI,
            sourceId: meta.doi,
            sourceTitle: meta.title,
            sourceAuthors: authors.isBlank ? nil : authors,
            sourceAbstract: meta.abstract,
            sourceExtra: extra
        )
    }

    // MARK: - URL extraction

    /// Port of `extractAllUrls(text)`. Matches full URLs including query strings (`?k=v`) and
    /// fragments (`#anchor`); trailing punctuation that is syntactically part of the surrounding
    /// sentence is stripped after extraction. CONVENTIONS §10: the trailing-punctuation strip is a
    /// `while`-loop over `.,)]!?;` (not `trimmingCharacters`, which trims both ends).
    private func extractAllUrls(_ text: String) -> [String] {
        var results: [String] = []
        for match in text.matches(of: Self.urlRegex) {
            var value = String(match.0)
            let trailing: Set<Character> = [".", ",", ")", "]", "!", "?", ";"]
            while let last = value.last, trailing.contains(last) {
                value.removeLast()
            }
            results.append(value)
        }
        return results
    }

    // MARK: - arXiv

    private func resolveArxiv(_ url: String) async -> SourceInfo? {
        guard let match = url.firstMatch(of: ArxivClient.arxivIdRegex) else { return nil }
        let id = String(match.1).replacing(Self.versionSuffixRegex, with: "")

        let meta = await withRetry { await arxivClient.fetchPaper(id) }
        guard let meta else { return nil }
        let extra = buildExtraJson([
            ("published", meta.published),
            ("categories", meta.categories)
        ])
        return SourceInfo(
            sourceType: .ARXIV,
            sourceId: meta.id,
            sourceTitle: meta.title,
            sourceAuthors: meta.authors.joined(separator: ", "),
            sourceAbstract: meta.abstract,
            sourceExtra: extra
        )
    }

    /// HuggingFace papers links embed arXiv IDs. Port of `resolveHfPaper(url)`.
    private func resolveHfPaper(_ url: String) async -> SourceInfo? {
        guard let match = url.firstMatch(of: Self.hfPaperRegex) else { return nil }
        let id = String(match.1)
        return await resolveArxiv("https://arxiv.org/abs/\(id)")
    }

    // MARK: - GitHub

    private func resolveGithub(_ url: String) async -> SourceInfo? {
        guard let match = url.firstMatch(of: Self.githubRegex) else { return nil }
        let owner = String(match.1)
        var repo = String(match.2)
        if repo.hasSuffix(".git") { repo.removeLast(4) }
        if owner.isBlank || repo.isBlank { return nil }

        let r = await withRetry { await githubApi.getRepo(owner: owner, repo: repo) }
        guard let r else { return nil }
        let extra = buildExtraJson([
            ("stars", r.stars),
            ("language", r.language),
            ("topics", r.topics),
            ("pushedAt", r.pushedAt)
        ])
        return SourceInfo(
            sourceType: .GITHUB,
            sourceId: r.fullName,
            sourceTitle: r.fullName,
            sourceAuthors: r.owner.login,
            sourceAbstract: r.description,
            sourceExtra: extra
        )
    }

    // MARK: - HuggingFace

    private func resolveHuggingFace(_ url: String) async -> SourceInfo? {
        // Exclude known non-model paths.
        if url.contains("/spaces/") || url.contains("/papers/") || url.contains("/docs/") {
            return nil
        }
        if let datasetMatch = url.firstMatch(of: Self.hfDatasetRegex) {
            return await resolveHfDataset(String(datasetMatch.1))
        }
        if let modelMatch = url.firstMatch(of: Self.hfModelRegex) {
            return await resolveHfModel(String(modelMatch.1))
        }
        return nil
    }

    private func resolveHfModel(_ id: String) async -> SourceInfo? {
        let r = await withRetry { await huggingFaceApi.getModel(id) }
        guard let r else { return nil }
        let modelId = r.modelId ?? r.id ?? id
        let extra = buildExtraJson([
            ("downloads", r.downloads),
            ("likes", r.likes),
            ("pipelineTag", r.pipelineTag),
            ("tags", Array(r.tags.prefix(8)))
        ])
        return SourceInfo(
            sourceType: .HUGGING_FACE,
            sourceId: modelId,
            sourceTitle: modelId,
            sourceAuthors: r.author,
            sourceAbstract: r.description.map { String($0.prefix(500)) },
            sourceExtra: extra
        )
    }

    private func resolveHfDataset(_ id: String) async -> SourceInfo? {
        let r = await withRetry { await huggingFaceApi.getDataset(id) }
        guard let r else { return nil }
        let datasetId = r.id ?? id
        let extra = buildExtraJson([
            ("downloads", r.downloads),
            ("likes", r.likes),
            ("tags", Array(r.tags.prefix(8)))
        ])
        return SourceInfo(
            sourceType: .HUGGING_FACE,
            sourceId: datasetId,
            sourceTitle: datasetId,
            sourceAuthors: r.author,
            sourceAbstract: r.description.map { String($0.prefix(500)) },
            sourceExtra: extra
        )
    }

    // MARK: - Extra JSON

    /// Port of `buildExtraJson(data)`. Drops nil values, serializes lists as JSON arrays, everything
    /// else as its natural JSON value. Mirrors `org.json.JSONObject` insertion order via an ordered
    /// pair list (Swift `Dictionary` is unordered; the Kotlin `mapOf` preserved insertion order and
    /// `JSONObject` echoes it, so we serialize manually to keep key order stable).
    private func buildExtraJson(_ data: [(String, Any?)]) -> String {
        var obj: [String: Any] = [:]
        var orderedKeys: [String] = []
        for (k, v) in data {
            guard let v else { continue }
            if let list = v as? [String] {
                obj[k] = list
            } else {
                obj[k] = v
            }
            orderedKeys.append(k)
        }
        return Self.serializeOrdered(obj, order: orderedKeys)
    }

    /// Serializes a small flat object preserving `order`, escaping strings the way `org.json` does for
    /// the value kinds we emit (strings, ints, string arrays).
    private static func serializeOrdered(_ obj: [String: Any], order: [String]) -> String {
        var parts: [String] = []
        for key in order {
            guard let value = obj[key] else { continue }
            parts.append("\(jsonString(key)):\(jsonValue(value))")
        }
        return "{" + parts.joined(separator: ",") + "}"
    }

    private static func jsonValue(_ value: Any) -> String {
        switch value {
        case let s as String:
            return jsonString(s)
        case let arr as [String]:
            return "[" + arr.map { jsonString($0) }.joined(separator: ",") + "]"
        case let i as Int:
            return String(i)
        case let i as Int64:
            return String(i)
        case let d as Double:
            return String(d)
        case let b as Bool:
            return b ? "true" : "false"
        default:
            return jsonString("\(value)")
        }
    }

    /// JSON-escapes a string per RFC 8259 (mirrors `org.json` string quoting for the chars we emit).
    private static func jsonString(_ s: String) -> String {
        var out = "\""
        for scalar in s.unicodeScalars {
            switch scalar {
            case "\"": out += "\\\""
            case "\\": out += "\\\\"
            case "\n": out += "\\n"
            case "\r": out += "\\r"
            case "\t": out += "\\t"
            case "\u{08}": out += "\\b"
            case "\u{0C}": out += "\\f"
            default:
                if scalar.value < 0x20 {
                    out += String(format: "\\u%04x", scalar.value)
                } else {
                    out.unicodeScalars.append(scalar)
                }
            }
        }
        out += "\""
        return out
    }

    // MARK: - Regexes (byte-identical to the Kotlin patterns)

    /// `https?://[^\s<>"']+`
    private nonisolated(unsafe) static let urlRegex = /https?:\/\/[^\s<>"']+/
    /// `v\d+$`
    private nonisolated(unsafe) static let versionSuffixRegex = /v\d+$/
    /// `huggingface\.co/papers/([\d.]+)`
    private nonisolated(unsafe) static let hfPaperRegex = /huggingface\.co\/papers\/([\d.]+)/
    /// `github\.com/([a-zA-Z0-9_.-]+)/([a-zA-Z0-9_.-]+)`
    private nonisolated(unsafe) static let githubRegex = /github\.com\/([a-zA-Z0-9_.-]+)\/([a-zA-Z0-9_.-]+)/
    /// `huggingface\.co/datasets/([a-zA-Z0-9_.-]+/[a-zA-Z0-9_.-]+)`
    private nonisolated(unsafe) static let hfDatasetRegex = /huggingface\.co\/datasets\/([a-zA-Z0-9_.-]+\/[a-zA-Z0-9_.-]+)/
    /// `huggingface\.co/([a-zA-Z0-9_.-]+/[a-zA-Z0-9_.-]+)`
    private nonisolated(unsafe) static let hfModelRegex = /huggingface\.co\/([a-zA-Z0-9_.-]+\/[a-zA-Z0-9_.-]+)/
}
